### DISCUSSION TITLE
Linking gettext for Python on M1 Macs

### DISCUSSION THREAD

**Discussion by dnicolson** [Homebrew/discussions#3291] [Everyday usage] [answered]
I have noticed that some Python Autoconf scripts fail on M1 Macs but do not on Intel Macs when `gettext` is installed via Homebrew.

I see that `keg_only` was removed in https://github.com/Homebrew/homebrew-core/pull/53489, linking still appears to be needed on M1 Macs though.

Here is a reduced example:
```
cat > test.c <<EOF
#include <Python.h>

int main() {
  Py_Initialize();
}
EOF
```

Intel:
`gcc test.c $(python-config --cflags --ldflags --embed)`

M1:
`gcc test.c -L/opt/homebrew/opt/gettext/lib $(python-config --cflags --ldflags --embed)`

After the removal of Python on macOS 12.3+, it looks as though more builds will break because of the path change from `/usr/local`. Is there any clean solution to this problem?

**Comment by gromgit** [comment]
What's the output of `which -a python-config`?

Also, note that neither of your command lines actually reference anything in `gettext`. If they did, you'd likely find that even the Intel one fails:
```
$ gcc test-python.c $(python3-config --cflags --ldflags --embed) -lintl
ld: library not found for -lintl
clang: error: linker command failed with exit code 1 (use -v to see invocation)

$ gcc test-python.c -L/usr/local/lib $(python3-config --cflags --ldflags --embed) -lintl

$ otool -L a.out
a.out:
	/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/Python (compatibility version 3.9.0, current version 3.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1858.112.0)
	/usr/local/opt/gettext/lib/libintl.8.dylib (compatibility version 11.0.0, current version 11.0.0)
```

As a general rule of thumb, building non-Homebrew software with Homebrew dependencies works best if `opt/lib` paths are explicitly provided during build configuration. That way, builds always succeeded even if dependencies are `keg_only`, regardless of platform.

**Comment by dnicolson** [comment]
Sorry, I neglected to mention Python is installed via asdf/pyenv. I am also now unable to reproduce the initial issue using only Homebrew. There appear to be two issues that are unrelated to Homebrew, sorry for the trouble!

```
which -a python-config
/Users/dave/.asdf/shims/python-config
```

The commands do work on Intel and M1 respectively (albeit with `was built for newer macOS version (12.3) than being linked (12.0)` warnings). If `-L/opt/homebrew/opt/gettext/lib` is omitted on an M1 Mac the following is output:
`ld: library not found for -lintl`

Interestingly, removing the [Homebrew search paths](https://github.com/native-api/pyenv/blob/3f7d5fcff81f9115872470dec661be14642a6cdf/plugins/python-build/bin/python-build#L1331-L1342) in pyenv seems to resolve the issue.

This works:
```
otool -L ~/.pyenv/versions/3.10.3/bin/python
/Users/dave/.pyenv/versions/3.10.3/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1858.112.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```

This doesn't:
```
otool -L ~/.pyenv/versions/3.10.3/bin/python
/Users/dave/.pyenv/versions/3.10.3/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1858.112.0)
	/opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 11.0.0, current version 11.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```

**Comment by dnicolson** [comment, marked as answer]
Related discussion: https://github.com/Homebrew/discussions/discussions/3321
pyenv workaround: https://github.com/pyenv/pyenv/commit/bc13a87beebfafb0e1baa352c0ba396fae4754b0
